### ISSUE TITLE
Multi-core server processors adaptation

### ISSUE THREAD

**Issue by sjsoft2015** [KeenSoftwareHouse/SpaceEngineers#20]
Maybe it's time to make the dedicated server more friendly to multi-core server processors?

**Comment by Drakoskai** [comment]
I'm playing around with that right no

**Comment by Spartan322** [comment]
be careful, multithreading is a bitch sometimes.

**Comment by Drakoskai** [comment]
I do it for 40 hours a week I should be ok

**Comment by Spartan322** [comment]
ok, cool

**Comment by Drakoskai** [comment]
I'm playing around with some ideas.  

I've got some models I'm introducing into codebase, trying to adapt the current threading code, ect

I've worked with these patterns in Java https://github.com/LMAX-Exchange/disruptor/tree/master/src/main/java/com/lmax/disruptor

https://github.com/JCTools/JCTools

I've done some research and the CSharp memory model is very similar and the patterns will apply.  
Also I took some inspiration from idsoftware's 
https://github.com/id-Software/DOOM-3-BFG

Specifically: http://fabiensanglard.net/doom3_bfg/threading.php

some more info I'm reading:
https://msdn.microsoft.com/en-ca/library/windows/desktop/ee416321(v=vs.85).aspx
"Beyond Programming shaders"
http://s09.idav.ucdavis.edu/talks/05-JP_id_Tech_5_Challenges.pdf

 Also an interview with John Carmack:
http://fabiensanglard.net/doom3/interviews.php

I have a ton more info but I think that's probably the most informative with out delving into CPU fencing and stuff.

/r

**Comment by sjsoft2015** [comment]
Good luck, pal, good luck!

**Comment by Drakoskai** [comment]
I'm going to start by building unit and performance tests for the current concurrency classes and seeing if Keen will accept them.  If that occurs, we can let the tests guide the direction of addressing this issue.  I'd prefer to test and measure and pinpoint the exact bottlenecks rather than sling code at the engine and hope for the best.

**Comment by Drakoskai** [comment]
I really hate the github interface... I need to setup the tortoise-git client....

**Comment by Drakoskai** [comment]
Test Specifications:
Let P be the number of producers, T be the number of threads and C be the number of consumers

OneToOneThroughput
P = 1, T= 4, C = 1.  A single producer produces N jobs that will be handled by T threads and consumed by 1 consumer.

OneToThreeDiamond
P = 1, T= 4, C = 1.  A single producer produces 2 jobs that will be handled by T threads and will feed into a 3rd job which is consumed by 1 consumer.

ThreeToOneThroughput
P = 3, T = 4, C = 1.  Three producers produce N jobs which will be handled by T threads and consumed by 1 consumer.
 
*All jobs must execute in order and on separate threads

TODO: 

1) Test decomposition of a large task into multiple tasks to be executed in parallel.  Example: Split a job to transform a large collection of elements into smaller collections and spawn jobs to transform the smaller collections on separate threads in order.

2) Assert CPU load balancing

**Comment by Drakoskai** [comment]
This issue needs some more love!

**Comment by Drakoskai** [comment]
Good news everyone! first round of test data is in.

These are the results of the stacking the Parallel class up against a similar .net Vanilla impl.  As you can see, Keen's  impl kicks crap out of  the standard one's garbage collection. However, the standard impl does way more calcs in way less time.

This test simply adds two values 2 million times.
 
<html>
	<head>
		<title>Test Report</title>
	</head>
	<body>
        Local time: 5/20/2015 11:12:07 PM<br>
        UTC time: 5/21/2015 3:12:07 AM
        <h2>Host configuration</h2>
Name: ULTIMA-PC<br>
Manufacturer: ASUS<br>
Model: All Series<br>
<br>
Operating System: Microsoft Windows 7 Ultimate <br>
 - Version: 6.1.7601<br>
 - ServicePack: 1<br>
<br>
Number of Processors: 1<br>
 - Name: Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz<br>
 - Description: Intel64 Family 6 Model 60 Stepping 3<br>
 - ClockSpeed: 3501 Mhz
<br>
 - Number of cores: 4<br>
 - Number of logical processors: 8<br>
 - Hyperthreading: ON<br>
<br>
Memory: 16321 MBytes<br>
L1Cache: 256 KBytes<br>
L2Cache: 1024 KBytes<br>
L3Cache: 8192 KBytes<br>
        <b><font color='red'>Hyperthreading is enabled.</font></b><br>
        <h2>Test configuration</h2>
        Runs: 3<br>
        <h2>Test results</h2>
        Best results of 3 run(s).<br>
        <br>
        <h2>Detailed test results</h2>
        <table border="1">
            <tr>
        <td>Scenarios</td>
       <td>Implementation</td>
                <td>Run</td>
                <td>Operations per second</td>
                <td>Duration (ms)</td>
                <td># GC (0-1-2)</td>
            </tr>
            <tr>
                <td>OneToOne</td>
                <td>StandardThroughputAsBaseLine</td>
                <td>0</td>
                <td>  684 228ops</td>
                <td>2923(ms)</td>
                <td>115-30-14</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>OneToOne</td>
                <td>StandardThroughputAsBaseLine</td>
                <td>1</td>
                <td>  787 401ops</td>
                <td>2540(ms)</td>
                <td>113-27-14</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>OneToOne</td>
                <td>StandardThroughputAsBaseLine</td>
                <td>2</td>
                <td>  864 677ops</td>
                <td>2313(ms)</td>
                <td>112-29-14</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>OneToOne</td>
                <td>KeenThroughput</td>
                <td>0</td>
                <td>  107 026ops</td>
                <td>18688(ms)</td>
                <td>5-3-1</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>OneToOne</td>
                <td>KeenThroughput</td>
                <td>1</td>
                <td>  107 204ops</td>
                <td>18656(ms)</td>
                <td>4-2-1</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>OneToOne</td>
                <td>KeenThroughput</td>
                <td>2</td>
                <td>  107 072ops</td>
                <td>18679(ms)</td>
                <td>5-3-1</td>
                <td>&nbsp;</td>
            </tr>
        </table>

**Comment by Drakoskai** [comment]
As another comparison, this is another 3rd party concurrency library that's used in high frequency finance/trading

<html>
	<head>
		<title>Disruptor-net - Test Report</title>
	</head>
	<body>
        Local time: 5/16/2015 9:36:54 AM<br>
        UTC time: 5/16/2015 1:36:54 PM
        <h2>Host configuration</h2>
Name: ULTIMA-PC<br>
Manufacturer: ASUS<br>
Model: All Series<br>
<br>
Operating System: Microsoft Windows 7 Ultimate <br>
 - Version: 6.1.7601<br>
 - ServicePack: 1<br>
<br>
Number of Processors: 1<br>
 - Name: Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz<br>
 - Description: Intel64 Family 6 Model 60 Stepping 3<br>
 - ClockSpeed: 3501 Mhz
<br>
 - Number of cores: 4<br>
 - Number of logical processors: 8<br>
 - Hyperthreading: ON<br>
<br>
Memory: 16321 MBytes<br>
L1Cache: 256 KBytes<br>
L2Cache: 1024 KBytes<br>
L3Cache: 8192 KBytes<br>
        <b><font color='red'>Hyperthreading can degrade performance, you should turn it off.</font></b><br>
        <h2>Test configuration</h2>
        Scenarios: All<br>
        Implementations: All<br>
        Runs: 3<br>
        <h2>Test results</h2>
        Best results of 3 run(s).<br>
        <br>
        <h2>Detailed test results</h2>
        <table border="1">
            <tr>
                <td>Scenario</td>
                <td>Implementation</td>
                <td>Run</td>
                <td>Operations per second</td>
                <td>Duration (ms)</td>
                <td># GC (0-1-2)</td>
                <td>Comments
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>Disruptor</td>
                <td>0</td>
                <td> 15 144 631ops</td>
                <td>7001(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>Disruptor</td>
                <td>1</td>
                <td> 14 148 273ops</td>
                <td>7405(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>Disruptor</td>
                <td>2</td>
                <td> 15 204 500ops</td>
                <td>6914(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>DisruptorWithAffinity</td>
                <td>0</td>
                <td> 11 140 819ops</td>
                <td>9314(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>DisruptorWithAffinity</td>
                <td>1</td>
                <td> 11 235 955ops</td>
                <td>9292(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>DisruptorWithAffinity</td>
                <td>2</td>
                <td> 11 215 791ops</td>
                <td>9252(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>BlockingCollection</td>
                <td>0</td>
                <td> 2 998 500ops</td>
                <td>675(ms)</td>
                <td>4-2-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>BlockingCollection</td>
                <td>1</td>
                <td> 2 793 296ops</td>
                <td>723(ms)</td>
                <td>4-1-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1C</td>
                <td>BlockingCollection</td>
                <td>2</td>
                <td> 4 273 504ops</td>
                <td>475(ms)</td>
                <td>3-1-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>Disruptor</td>
                <td>0</td>
                <td> 13 135 426ops</td>
                <td>9653(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>Disruptor</td>
                <td>1</td>
                <td> 12 817 226ops</td>
                <td>9845(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>Disruptor</td>
                <td>2</td>
                <td> 12 363 996ops</td>
                <td>10153(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>DisruptorWithAffinity</td>
                <td>0</td>
                <td> 10 844 810ops</td>
                <td>11258(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>DisruptorWithAffinity</td>
                <td>1</td>
                <td> 10 559 662ops</td>
                <td>11535(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>DisruptorWithAffinity</td>
                <td>2</td>
                <td> 10 820 168ops</td>
                <td>11306(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>BlockingCollection</td>
                <td>0</td>
                <td> 1 164 144ops</td>
                <td>882(ms)</td>
                <td>7-4-1</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>BlockingCollection</td>
                <td>1</td>
                <td> 1 114 827ops</td>
                <td>919(ms)</td>
                <td>7-4-1</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>MultiCast1P3C</td>
                <td>BlockingCollection</td>
                <td>2</td>
                <td> 1 137 656ops</td>
                <td>900(ms)</td>
                <td>8-4-1</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>Disruptor</td>
                <td>0</td>
                <td> 13 419 216ops</td>
                <td>8010(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>Disruptor</td>
                <td>1</td>
                <td> 12 525 050ops</td>
                <td>8545(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>Disruptor</td>
                <td>2</td>
                <td> 10 305 028ops</td>
                <td>10263(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>DisruptorWithAffinity</td>
                <td>0</td>
                <td> 9 275 577ops</td>
                <td>11332(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>DisruptorWithAffinity</td>
                <td>1</td>
                <td> 9 848 335ops</td>
                <td>10704(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>DisruptorWithAffinity</td>
                <td>2</td>
                <td> 9 477 774ops</td>
                <td>11102(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>BlockingCollection</td>
                <td>0</td>
                <td> 1 338 688ops</td>
                <td>754(ms)</td>
                <td>12-5-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>BlockingCollection</td>
                <td>1</td>
                <td> 1 200 480ops</td>
                <td>841(ms)</td>
                <td>13-5-1</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3Step</td>
                <td>BlockingCollection</td>
                <td>2</td>
                <td> 1 414 427ops</td>
                <td>713(ms)</td>
                <td>13-5-1</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>Disruptor</td>
                <td>0</td>
                <td> 13 366 005ops</td>
                <td>4490(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>Disruptor</td>
                <td>1</td>
                <td> 13 131 976ops</td>
                <td>4569(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>Disruptor</td>
                <td>2</td>
                <td> 13 345 195ops</td>
                <td>4496(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>DisruptorWithAffinity</td>
                <td>0</td>
                <td> 11 474 469ops</td>
                <td>5231(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>DisruptorWithAffinity</td>
                <td>1</td>
                <td> 12 190 166ops</td>
                <td>4923(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>DisruptorWithAffinity</td>
                <td>2</td>
                <td> 11 409 013ops</td>
                <td>5260(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>BlockingCollection</td>
                <td>0</td>
                <td> 7 425 742ops</td>
                <td>405(ms)</td>
                <td>4-2-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>BlockingCollection</td>
                <td>1</td>
                <td> 7 672 634ops</td>
                <td>391(ms)</td>
                <td>2-1-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Sequencer3P1C</td>
                <td>BlockingCollection</td>
                <td>2</td>
                <td> 6 912 442ops</td>
                <td>435(ms)</td>
                <td>2-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>Disruptor</td>
                <td>0</td>
                <td> 12 434 717ops</td>
                <td>8779(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>Disruptor</td>
                <td>1</td>
                <td> 12 960 082ops</td>
                <td>8442(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>Disruptor</td>
                <td>2</td>
                <td> 11 824 524ops</td>
                <td>9182(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>DisruptorWithAffinity</td>
                <td>0</td>
                <td> 9 450 902ops</td>
                <td>11313(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>DisruptorWithAffinity</td>
                <td>1</td>
                <td> 9 432 182ops</td>
                <td>11328(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>DisruptorWithAffinity</td>
                <td>2</td>
                <td> 9 843 488ops</td>
                <td>10883(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>BlockingCollection</td>
                <td>0</td>
                <td> 1 186 239ops</td>
                <td>852(ms)</td>
                <td>5-2-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>BlockingCollection</td>
                <td>1</td>
                <td> 1 353 179ops</td>
                <td>746(ms)</td>
                <td>6-2-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>DiamondPath1P3C</td>
                <td>BlockingCollection</td>
                <td>2</td>
                <td> 1 386 962ops</td>
                <td>728(ms)</td>
                <td>5-1-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>Disruptor</td>
                <td>0</td>
                <td>mean=176.15ns, 99%=128ns, 99.99%=262144</td>
                <td>2341(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>Disruptor</td>
                <td>1</td>
                <td>mean=129.58ns, 99%=128ns, 99.99%=32768</td>
                <td>2346(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>Disruptor</td>
                <td>2</td>
                <td>mean=132.52ns, 99%=128ns, 99.99%=16384</td>
                <td>2340(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>DisruptorWithAffinity</td>
                <td>0</td>
                <td>mean=17529.08ns, 99%=524288ns, 99.99%=8388608</td>
                <td>2388(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>DisruptorWithAffinity</td>
                <td>1</td>
                <td>mean=63648.09ns, 99%=4194304ns, 99.99%=16777216</td>
                <td>2385(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>DisruptorWithAffinity</td>
                <td>2</td>
                <td>mean=1147.36ns, 99%=16384ns, 99.99%=524288</td>
                <td>2429(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>BlockingCollection</td>
                <td>0</td>
                <td>mean=1236.37ns, 99%=8192ns, 99.99%=262144</td>
                <td>1506(ms)</td>
                <td>6-3-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>BlockingCollection</td>
                <td>1</td>
                <td>mean=4718.41ns, 99%=32768ns, 99.99%=2097152</td>
                <td>1501(ms)</td>
                <td>6-2-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>Pipeline3StepLatency</td>
                <td>BlockingCollection</td>
                <td>2</td>
                <td>mean=8872.49ns, 99%=262144ns, 99.99%=2097152</td>
                <td>1502(ms)</td>
                <td>6-2-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>Disruptor</td>
                <td>0</td>
                <td> 50 075 112ops</td>
                <td>2444(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>Disruptor</td>
                <td>1</td>
                <td> 48 076 923ops</td>
                <td>2425(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>Disruptor</td>
                <td>2</td>
                <td> 47 984 644ops</td>
                <td>2423(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>DisruptorWithAffinity</td>
                <td>0</td>
                <td> 21 141 649ops</td>
                <td>5070(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>DisruptorWithAffinity</td>
                <td>1</td>
                <td> 21 074 815ops</td>
                <td>5097(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>DisruptorWithAffinity</td>
                <td>2</td>
                <td> 20 379 050ops</td>
                <td>5269(ms)</td>
                <td>0-0-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>BlockingCollection</td>
                <td>0</td>
                <td> 2 898 550ops</td>
                <td>697(ms)</td>
                <td>4-2-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>BlockingCollection</td>
                <td>1</td>
                <td> 3 210 272ops</td>
                <td>629(ms)</td>
                <td>4-1-0</td>
                <td>&nbsp;</td>
            </tr>
            <tr>
                <td>UniCast1P1CBatch</td>
                <td>BlockingCollection</td>
                <td>2</td>
                <td> 3 384 094ops</td>
                <td>598(ms)</td>
                <td>4-1-0</td>
                <td>&nbsp;</td>
            </tr>
        </table>

**Comment by Spartan322** [comment]
I'm not sure how to respond seeing as concurrency libraries may be different for different purposes and reasons, if these were all from just logic calculations (adding to an object 2 million times) then Id discount it because they are testing the areas the libraries perform best in.

**Comment by Drakoskai** [comment]
Well, we're not done yet, as I've stated. This issue is actually poorly defined. However it begs the question of "is the concurrency adequate".  Testing will prove yes or no.  After testing we may find that the keen implementations are fine the way they are, or we may find ways to improve them.  If we start cowboy coding and just rewriting code to see what happens we won't really know if the changes are necessary or even useful.  

So far we've shown that Keen's code is good at not making garbage and not so good at executing rapid tasking.  I haven't drawn a conclusion that keen's code needs to support rapid tasking efficiently, to be honest I'm trying not to pass any judgement at all -- all we've done is measure it.

On to the next test!

One of the things I would like to do is load a large voxel set and perform a large-scale transform.  If anyone has some ideas for other ways to measure that would be helpful.

**Comment by Drakoskai** [comment]
Can we get someone to close this issue? 

The results are that SE does support multi threading, but individual jobs may not be optimized to be broken down and executed.  This issue should be approached on a job by job basis rather than from the perspective of "re-doing" all the concurrency.  As it stands, the current algorithms add a lot of latency from context switching using un-optimized locking -- but we could optimize this forever and see no improvement if the jobs themselves are trying to execute a rather large series of calculations or logic on a single worker as opposed to a job starting on a worker thread and then submitting sub-jobs to be worked on by other worker threads.  

The benchmarks above show that SE's concurrency's latency comes from every time a new job is created and that job ramps up does it's work and then calls back.  So great care needs to be taken to make sure that the job's complexity warrants the overhead that will be added and will still execute faster than if it was done on a single thread.  

Final note: Also blocking caused by locking should be considered too, since when a worker is locking a resource all other threads must wait while they are blocked.  This would create a sort of "artificial" latency that would be hard to measure because of the difficulty of producing this effect with all the components integrated like they would be during normal game play.

TL;DR
Create issues on the jobs rather than from the perspective of the whole concurrency system and close this one.

**Comment by zrisher** [comment]
Thanks for all your work on this, it's way over my head but I'm really glad there's work being done on it.

**Comment by joemorin73** [comment]
Because the threads have a cost on creation and destruction, is there a way to not need create/destroy them?  IE, recycle threads or threads wait for tasks?  This might increase the performance.

**Comment by Drakoskai** [comment]
No there isn't because it's not about creation and destruction, its a matter of context switching, and locking and blocking.   The threads are already in a thread pool and are being re-used.

**Comment by joemorin73** [comment]
IE, negligible improvement.  I get this won't be easy to be done.  Is this something Keen stated they were working on?

**Comment by Drakoskai** [comment]
I could re-write their concurrency, but it would be a drastic change and keen would be unlikely to accept it.  It would also take a large number of hours and I have a full time programming job as it is.  So I don't feel like the effort would be worth it since it would likely be wasted anyway.  It would be better to optimize on a case by case basis since there is probably room for improvement on the way the individual tasks are being executed. For example: it looks like their "Sovereign" threads, the ones that run in the background, are probably trying to execute some their own calculations on their own thread.  This may be optimal with their current multi-threading libraries, but it also may not be.

**Comment by sjsoft2015** [comment]
Very sorry to hear that, though I had such concerns. At least have the opportunity to profile an application and find out at what point is the load on the server? Display the data to the server admins in real-time so that they can quickly remove the load?

**Comment by Drakoskai** [comment]
I'll take a look this week into some of the voxel calc jobs.  
This
https://github.com/KeenSoftwareHouse/SpaceEngineers/blob/master/Sources/Sandbox.Game/Engine/Voxels/MyPrecalcJobPhysicsBatch.cs
is basically the equivalent of a "Future"
http://en.wikipedia.org/wiki/Futures_and_promises

And that's where I will probably start looking.  I'd like to see how many calculations that job will attempt to perform on a single thread.

**Comment by Drakoskai** [comment]
Also we might want to build a unit performance test around https://github.com/KeenSoftwareHouse/SpaceEngineers/blob/master/Sources/Sandbox.Game/Engine/Voxels/MyOctreeStorage.cs
Maybe there needs to be a small thread pool devoted to voxel work.  there's two methods in there, WriteRange and ReadRange, maybe they need to be multi-threaded jobs if the write or read is > some amount of voxels.  This is all speculation at this point.

**Comment by Drakoskai** [comment]
Last night I added some padding to the WorkItem to keep it out of the cores' local cache and in the shared cache.

I guess I should stop for a second and explain a bit...

Cache misses/false sharing: 
    Basically a processor core will do work in it's local cache in typically chucks of 64 bytes, so you basically get whatever is on that line.  Well, with OOP the way it is, Objects are composed of references so their members are stored elsewhere.  A Processor core will try to store all the data on the things they are working with in their L1 cache because it's the fastest and closest to the core. A cache will pull data to it in 64 byte chunks which is usually referred to as a cache page and page is composed of 8 byte lines, a long/Int64 will fit nicely on 1 line.  Iterating over an array of longs is super efficient.  They will all be next to each other and life will be great.  The problem is that items in a typical data structure can't sit next to each other because they are too big.  Every element you iterate over will be a cache miss, and you'll have to go looking for it in the main memory to pull it down to the cache. Once this is all done, the results will have to be pushed back up and out of the L1 cache because well, it's not shared.  At this point typical algorithms will stop the world via a lock and push it up to prevent a data race.  If you happen to get other variables on the page also you will create contention over them with other cores too and they will then lag as well.  Cache padding adds buffer data to the item so that it doesn't fit in the L1 cache this means no contention and no cache misses and allows for you to forgo a lock that will use a harsher memory fence in favor atomic CompareAndSwap operations to perform at max efficiency with a lighter memory fence.  

Basically all the Big O optimization in the world will mean squat in the face of this issue.

back on topic...
I saw a performance increase up to 120,000 -ish operations per second with 16000 ms to perform the entire 2 million add value routines. Still not good enough, but it was a minor tweak that brought down the cost of the doing something on another thread slightly.  I think If I change some of the locks to CAS operations we'll probably start seeing some major gains. But it's still just peanuts honestly -- there is some major work to be done here and testing.

(btw each of the 2 million add value operations are done on separate threads to measure the cost of switching threads, not doing the work on the thread.  Lower thread switching latency = smaller jobs = better multi-core load balancing) 

Wish I had more time to work on this

**Comment by Drakoskai** [comment]
Last night I added some padding to the WorkItem to keep it out of the cores' local cache and in the shared cache.

I guess I should stop for a second and explain a bit...

Cache misses/false sharing: 
    Basically a processor core will do work in it's local cache in typically chucks of 64 bytes, so you basically get whatever is on that line.  Well, with OOP the way it is, Objects are composed of references so their members are stored elsewhere.  A Processor core will try to store all the data on the things they are working with in their L1 cache because it's the fastest and closest to the core. A cache will pull data to it in 64 byte chunks which is usually referred to as a cache page and page is composed of 8 byte lines, a long/Int64 will fit nicely on 1 line.  Iterating over an array of longs is super efficient.  They will all be next to each other and life will be great.  The problem is that items in a typical data structure can't sit next to each other because they are too big.  Every element you iterate over will be a cache miss, and you'll have to go looking for it in the main memory to pull it down to the cache along with other values near it. Once this is all done, the results will have to be pushed back up and out of the L1 cache. -- if another core operates on a variable on that cache line, the entire line will be invalidated, even other values that just happen to be near there, in all memory locations!   Padding makes the value take up a whole line so that when it's being passed around it doesn't drag anything with it and nothing else will invalidate it.

http://igoro.com/archive/gallery-of-processor-cache-effects/

Basically all the Big O optimization in the world will mean squat in the face of this issue.

back on topic...
I saw a performance increase up to 120,000 -ish operations per second with 16000 ms to perform the entire 2 million add value routines. Still not good enough, but it was a minor tweak that brought down the cost of the doing something on another thread slightly.  I think If I change some of the locks to CAS operations we'll probably start seeing some major gains. But it's still just peanuts honestly -- there is some major work to be done here and testing.

(btw each of the 2 million add value operations are done on separate threads to measure the cost of switching threads, not doing the work on the thread.  Lower thread switching latency = smaller jobs = better multi-core load balancing) 

Wish I had more time to work on this

**Comment by Drakoskai** [comment]
To anyone still paying attention to this thread: Ok, so I spent the last few weeks writing a new "parallel" library. I think I am seeing some pretty decent performance gains.  Right now my fork is basically a hybrid of the two systems.  It's been a lot of work.  I'm not sure how many memory leaks, or bugs I've introduced but I'd appreciate some help if anyone is interested. To be honest, I'm a Java dev not a C# one so I don't know how good/bad the code I wrote was.  I tried to translate my general compsci+ java knowledge into C# as best I could.

Gonna take a break and then rebury my head in it.

**Comment by Drakoskai** [comment]
I've got more integration work to do.  I'm seeing some issues where the caller is expecting to wait for the completion of the job and I'm not really facilitating that at the moment.

**Comment by zrisher** [comment]
Definitely still reading this thread, this would be amazing. Thanks for all your hard work.

**Comment by Drakoskai** [comment]
last night I was a little over tired and I broke it before the pull request,  so I closed it.  Heh that's what I get for coding for like 14 hours straight.  The sequence on the queue isn't advancing so it just waits for ever for the next sequence.  I take get it fixed in a few hours.

**Comment by Drakoskai** [comment]
fixt :)

**Comment by thelastWallE** [comment]
Is this only working on servers? How do we test this with the github build? Only local?

**Comment by Drakoskai** [comment]
right now I'm progressively debugging it so the only way to try it would be to download my fork, it will be a github only thing unless keen wants it

**Comment by Drakoskai** [comment]
btw, I've been using the Steam name Brutoss for awhile in multiplayer. I've switched to DrakosKai recently, so hit me up if you see me in game.

**Comment by Drakoskai** [comment]
It's working fairly well now, the hard part now is integrating it with KSH's Code.  There a bunch of different ways they are farming out tasks so need to look at that at this point.

**Comment by sjsoft2015** [comment]
you can do it! I think KSW should help.

**Comment by Drakoskai** [comment]
I'm currently working on the concept of batching.  For example, you want to queue up a bunch of tasks that will be farmed out to the first least loaded thread and executed, but you want them executed together.  I'm adding the ability to submit a batch of tasks.  Also I am look at adding Futures.  Keen already has a future class but it doesn't look like it's being used anywhere.  So I'm creating an IFuture Interface and then I will be backing it with an implementation that will be returned immediately after starting a task or batch of tasks and will allow the caller to track the progress or wait on the task to finish.

**Comment by Drakoskai** [comment]
I'm stealing a bunch of Code from Google's Guava for Java.  Pretty good api's in there.

**Comment by harag-on-steam** [comment]
Hm, I'm a Java guy, too, but I think it would be bad form to reinvent the wheel for something so basic. I'd be suprised if C# (what is the .NET version SE uses? 4.5?) had nothing to offer here. I quick search came up with [System.Threading.Tasks](https://msdn.microsoft.com/en-us/library/system.threading.tasks%28v=vs.110%29.aspx) and ["Parallel Programm with .NET"](https://msdn.microsoft.com/en-us/library/ff963553.aspx). There even seems to be [special syntax](https://msdn.microsoft.com/en-us/library/hh191443.aspx) for this.

**Comment by Drakoskai** [comment]
Just uploaded Futures + Fluent API for interacting with it.  I haven't fully tested it yet-- just wanted to get it up before heavensward launch and I become a hermit for a few days.

**Comment by krolli** [comment]
Hi,
this was an interesting read. It's just a shame I only found out about it now, otherwise I might have saved you some time. It seems you know a lot more about general multithreading and concurrency than I do, but you seem to struggle with what is going on in the SE code-base itself so I'll try to go over most of what I can think of off the top of my head. It is possible that you've already found out about these while you were digging around, but for the sake of completeness, here goes.

In general, the engine is built around 2 threads running all the time: update thread and render thread. Most things live in update thread and only DX9 and DX11 renderer is using the render thread. These two threads communicate using message queue (which is fairly complicated and not that interesting for this discussion). Apart from those two threads, there is certain number of workers that depends on the number of cores. We try to have the same number of threads as there are cores, but minimum is 3 (update, render, 1 worker) which means we get badly oversubscribed on dual core machines.

As for workers and ParallelTasks library you probably found in there, it was originally not ours, but we had to make changes to it (mostly fixing races and deadlocks IIRC). WorkStealingScheduler you've mentioned somewhere is actually not used anymore. We're currently using FixedPriorityScheduler so have a look at that if you haven't already. Since we're not actually using most of the stuff available in ParallelTasks, I was thinking of cleaning it up, removing unused parts and simplifying those that get used a lot. I doubt I will be able to do that anytime soon though. We're mostly using completion callbacks priorities that you can assign to tasks.

There are a few tasks that are short and latency intolerant (mostly coming from render thread and having to finish with the same frame) but most of the time is probably taken by voxel-related computations, especially with procedural asteroids, and those can sometimes be fairly fast (less than 1 ms) or incredibly slow (50-100 ms) and it is quite hard to know ahead of time which one it will be. You might notice there is a lot of extra classes related to voxel computation tasks (subclasses of MyPrecalcJob and queuing in MyPrecalcComponent). I'm not very happy with these adding extra complexity, but it was the best I could think of at the time, given the problems I was dealing with (extra worker causing update/render thread starvation on dual cores) and I'm not sure it still works as good as it used to, since the code has changed in various ways since then.

I hope this clears up anything you didn't understand and gives you somewhat better idea in terms of performance. Feel free to ask for more details, since I'm not sure whether you want to focus on server performance, where render-related computations don't exist (including render geometry for voxels) or general game performance. Sadly, those two can be closely related, since a lot of stuff is happening on both client and server.

As for improvements, I noticed you were adding some low level utilities. I personally wouldn't bother with AtomicBool, which looked like just a wrapper around Interlocked operations and some conversion utilities. Using Interlocked directly should be fine in most situations in my opinion. As for existing C# libraries that someone mentioned, they mostly have completely different goals, as your testing has most likely shown. We usually try to avoid allocations where possible, which seems to be very distant goal for .NET libraries.

So, sorry for the wall of text, there is really no TL;DR version. Feel free to ask anything if you're interested.

**Comment by Drakoskai** [comment]
"In general, the engine is built around 2 threads running all the time: update thread and render thread. Most things live in update thread and only DX9 and DX11 renderer is using the render thread. These two threads communicate using message queue (which is fairly complicated and not that interesting for this discussion). Apart from those two threads, there is certain number of workers that depends on the number of cores. We try to have the same number of threads as there are cores, but minimum is 3 (update, render, 1 worker) which means we get badly oversubscribed on dual core machines."

I saw that.  I didn't want to mess with the render thread or the update thread, I considered these threads to be sovereign.  My goal is to reduce the over head of using the worker threads.  Currently there's a lot of latency introduced by numerous issues that I talked about above.  Basically, given this you would want to figure out the complexity of a task so that it was complex enough to warrant putting it on another thread. So if you put a bunch of small tasks out there you would end up taking longer just putting it on a worker than you would have taken just executing it in the main thread.  This has the side effect of creating an imbalance on your processors since they are executing long tasks. 

So my goal was to reduce the overhead to allow for much smaller tasks and better load balance.

As for workers and ParallelTasks library you probably found in there, it was originally not ours, but we had to make changes to it (mostly fixing races and deadlocks IIRC). WorkStealingScheduler you've mentioned somewhere is actually not used anymore. We're currently using FixedPriorityScheduler so have a look at that if you haven't already. Since we're not actually using most of the stuff available in ParallelTasks, I was thinking of cleaning it up, removing unused parts and simplifying those that get used a lot. I doubt I will be able to do that anytime soon though. We're mostly using completion callbacks priorities that you can assign to tasks.

I found that blog ;p


"There are a few tasks that are short and latency intolerant (mostly coming from render thread and having to finish with the same frame) but most of the time is probably taken by voxel-related computations, especially with procedural asteroids, and those can sometimes be fairly fast (less than 1 ms) or incredibly slow (50-100 ms) and it is quite hard to know ahead of time which one it will be."

Consider breaking up the heavy task into a batch, the new library is supposed to keep all the tasks of a batch sequential despite being executed on different threads.  So you could break a large block of calculations up into smaller ones and group them as a batch -- the whole batch of tasks would be contained in one of the IFuture responses.

"You might notice there is a lot of extra classes related to voxel computation tasks (subclasses of MyPrecalcJob and queuing in MyPrecalcComponent). I'm not very happy with these adding extra complexity, but it was the best I could think of at the time, given the problems I was dealing with (extra worker causing update/render thread starvation on dual cores) and I'm not sure it still works as good as it used to, since the code has changed in various ways since then."

Yeah I was really interested in working with these two classes it seemed to me like this was a major bottleneck.

"I hope this clears up anything you didn't understand and gives you somewhat better idea in terms of performance. Feel free to ask for more details, since I'm not sure whether you want to focus on server performance, where render-related computations don't exist (including render geometry for voxels) or general game performance. Sadly, those two can be closely related, since a lot of stuff is happening on both client and server."

I'm not really clear on where the server starts and the client ends.  They seem very tightly coupled but I could be wrong.  From my understanding and, and I'm not sure on this, much of the server latency comes from syncing all the clients to the server rather than having all the logic done on the server and then just have the clients listening and sending player actions to the server.  I'm not sure though.


" As for existing C# libraries that someone mentioned, they mostly have completely different goals, as your testing has most likely shown. We usually try to avoid allocations where possible, which seems to be very distant goal for .NET libraries."

Yeah I know that you probably want to keep GC down and things like that.  I have classes that will eventually be object pools for some of the holder classes that I'm currently creating and destroying but I'm more worried about getting it 100% functioning correctly before I fully optimize it.

"So, sorry for the wall of text, there is really no TL;DR version. Feel free to ask anything if you're interested."

Totally cool, I'm just really glad you guys are interested!   Basically my concerns are just keeping it in line with what you guys want for the game. I'm trying to make a unified fluent API for it so it will be easy to work with something like:

future = Concurrent.CreateAsynchronousTask(inital task and that task's callback and exception handler)

future.Then(2nd task in the batch and it's callback and handler).Then(3rd task in the batch and it's callback and handler).Then( ect...)

future2 = future.After(1st task of a potential second batch).then(2nd task of 2nd batch).then(ect...)

future2.Now(); //Execute all chained futures

future2.Wait(); 

ect...

That's sort of the goal for the first pass on it.

/r

**Comment by krolli** [comment]
> I found that blog ;p

What blog? :P

> Consider breaking up the heavy task into a batch, the new library is supposed to keep all the tasks of a batch sequential despite being executed on different threads. So you could break a large block of calculations up into smaller ones and group them as a batch -- the whole batch of tasks would be contained in one of the IFuture responses.

Originally, voxel computations used to be quite small, but they were being spawned in thousands. Normally, that would be fine and I'd just focus on reducing memory overhead for each of those and leave them in the queue. Problem with voxels is that they are destructible. Even if writes and reads to voxel data are atomic, by splitting task into smaller ones you end up in a situation where the resulting geometry was not produced atomically and might be inconsistent. For render geometry, this can result in occasional cracks in mesh, which is annoying, but will get sorted out eventually. However, cracks in physics geometry can be disastrous, with player clipping through in the worst case.

Also, reading from the voxel storage is more efficient when done in larger batches, as opposed to several small ones, due to internal cost of traversing the octree structure of it. Breaking it up into smaller chunks might or might not improve performance due to the traversal and merging costs. All in all, I would say voxel computations are more about optimizing tasks themselves, rather than threading of those tasks. Besides, on clients you will usually see multiple computations in flight at any one time (typically I see about 20 of them queued up, hundreds if you're flying very fast) when levels of detail in render are being generated. On server, it's probably only physics mesh generation, but even there you should generally get 8+ tasks spawned at the same time.

> Yeah I was really interested in working with these two classes it seemed to me like this was a major bottleneck.

They are, but it's most likely not threading issue. They tend to push CPUs to 100% on clients so more smaller jobs are not very likely to help there in my opinion.  I'd be happy if you proved me wrong though. :) Maybe reducing overhead from various queues being used there would, since they are Deques wrapped in locks (not even using EnqueueFront anymore) which could be replaced with some non-blocking queues. Our data structures for threading are a mess, but less so than most of the ones in .NET standard library at least.

Also, the weird thing about them is they build another mechanism on top of normal scheduling, which is something I don't quite like. Getting rid of that would be nice.

> I'm not really clear on where the server starts and the client ends.

Right now, I'm not sure anyone knows. :P

> They seem very tightly coupled but I could be wrong. From my understanding and, and I'm not sure on this, much of the server latency comes from syncing all the clients to the server rather than having all the logic done on the server and then just have the clients listening and sending player actions to the server. I'm not sure though.

I believe you've got the gist of it. No-Brain would probably tell you more on that, but I think what you're describing is similar to what he has in mind for new network layer he's working on.

> Yeah I know that you probably want to keep GC down and things like that. I have classes that will eventually be object pools for some of the holder classes that I'm currently creating and destroying but I'm more worried about getting it 100% functioning correctly before I fully optimize it.

Sticking to value types where possible is probably better than pooling, especially if pools reuse everything and never release (peak in required item count would essentially turn into memory leak). But you're right with the optimization part.

> I'm trying to make a unified fluent API for it so it will be easy to work with something like ...

I admit futures are nice and all, but in most cases (in games) I haven't found that many uses for them, but maybe they were just not designed with that in mind. I can see them being useful if update loop was designed to be threaded, with clear dependency chains, but we're currently trapped in OOP model that is very threading unfriendly IMO. Having the ability to participate in scheduled tasks from main thread could help a lot. Not sure if this can be implemented as part of waiting on future (probably can, but it sounds like it might need some acrobatics), or  needs to be part of scheduler interface. Adding it to scheduler would be trivial (both support it already) but I'd be careful with it since one of the voxel tasks could block the participating thread for quite a while.

Also, I personally like the fixed priority approach to scheduling (at least in our case) and I was thinking of adding ability to keep at least one thread free for latency-intolerant short tasks, even if it could be working on low priority latency-tolerant stuff (voxels). Not quite as good as having reasonably sized tasks, so I'm not sure whether this is even worth pursuing (would also cause issues on dual cores :( )

**Comment by Drakoskai** [comment]
"I admit futures are nice and all, but in most cases (in games) I haven't found that many uses for them, but maybe they were just not designed with that in mind. I can see them being useful if update loop was designed to be threaded, with clear dependency chains, but we're currently trapped in OOP model that is very threading unfriendly IMO. Having the ability to participate in scheduled tasks from main thread could help a lot. Not sure if this can be implemented as part of waiting on future (probably can, but it sounds like it might need some acrobatics), or needs to be part of scheduler interface. Adding it to scheduler would be trivial (both support it already) but I'd be careful with it since one of the voxel tasks could block the participating thread for quite a while."

We don't have to wait on the future, I've seen that pattern in your code so I tried to replicate it.  But, you could use the future and just not wait on it.  The best solution, I think would be some kind of observable stream coming from the update loop that the render loop observes.  The trick would be getting into the update loop to make changes because players are doing things.  So this pattern that I've implemented uses the "single writer principle" to avoid locking while keep things in order.  So when you are farming tasks out, they will be executed in the order that they are added to the ring buffer due to the sequence tracking nature of it. The unit tests should test this feature.  So, if you want to put a change into the update loop, or millions of changes in a multi threaded context, as long as the update loop is paying attention to the mult-threaded portion, it will not have a race condition.  This should stay true even if a task spawns a task.  The tasks can wait or not wait based on their dependency needs.

If every activity is some kind of workflow like:

A player fires a weapon.

A weapon adds a projectile.

a projectile strikes an object and does damage.

Wouldn't this be just a workflow of workflows? A player fires a weapon is stateful, so internally that workflow needs to wait on any outside asycn calls but once the projectile is out in the world the logic doesn't need to wait any longer.  I guess what I'm trying to say is if your logic is stateful then you need some sort of waiting mechanism but not all things should be stateful.  I think all you really need to do is identify the stateful/not stateful parts and then do some minor tweaks so that you only wait on stateful stuff.  I think the news download  part of the main menu is a great place to test some of these things out because that class seems to have some very clear workflows.

"Also, I personally like the fixed priority approach to scheduling (at least in our case) and I was thinking of adding ability to keep at least one thread free for latency-intolerant short tasks, even if it could be working on low priority latency-tolerant stuff (voxels). Not quite as good as having reasonably sized tasks, so I'm not sure whether this is even worth pursuing (would also cause issues on dual cores :( )"

I've seen the priority queue and I'll see if I can incorporate it some how -- but the pattern that I've implemented ensures that there are no race conditions by keeping everything ordered -- priority would imply out of order tasking.  Thinking about it more, what happens in your current model if all threads are working on long running tasks and a priority task gets queued?

My mental image is basically the update loop has to take in all inputs and transform the entire world model in under 10 millis -- while not blocking the render loop to take a snapshot on it's own terms.  This seems normal, most application have some kind of stateful critical resource at their heart.  The trick is that nothing can block this and this can't block the render loop I think.  So I pictured the update loop as a set of asynchronous tasks.  The precalc/phisics job being the most beastly but there are many dependencies on it.

**Comment by krolli** [comment]
> If every activity is some kind of workflow like:
A player fires a weapon.
A weapon adds a projectile.
a projectile strikes an object and does damage.
Wouldn't this be just a workflow of workflows? A player fires a weapon is stateful, so internally that workflow needs to wait on any outside asycn calls but once the projectile is out in the world the logic doesn't need to wait any longer. I guess what I'm trying to say is if your logic is stateful then you need some sort of waiting mechanism but not all things should be stateful. I think all you really need to do is identify the stateful/not stateful parts and then do some minor tweaks so that you only wait on stateful stuff. I think the news download part of the main menu is a great place to test some of these things out because that class seems to have some very clear workflows.

I'm not sure I follow here, especially with the 'workflow' you mentioned. The example with player firing a projectile that strikes an object seems really odd. You create and entity that is added to simulation and that's it. Projectile striking an object and dealing damage is completely unrelated to firing a weapon, as there can be potentially tens or hundreds of frames in between. Handling of that is usually a result of reaction to collision event coming out of simulation. Trying to make a task for this would result in something that sits idle somewhere in the queue except at the beginning and end. And there would have to be a lot of additional stuff dealing with the case where nothing has been hit.

> Thinking about it more, what happens in your current model if all threads are working on long running tasks and a priority task gets queued?

Nothing good, which is another reason why I'm not happy with the way latency tolerant and intolerant stuff is interacting with each other on workers. It hasn't been a problem simply because all latency intolerant tasks are also being worked on by the issuing thread (render in most cases) so in the worst case it just isn't executed in parallel.

> So I pictured the update loop as a set of asynchronous tasks. The precalc/phisics job being the most beastly but there are many dependencies on it.

It is possible to think of it as a set of tasks, but you seem to lack the bigger picture when deciding what these tasks are. Those voxel geometry tasks are really somewhere on the edge and they barely matter to overall multithreading at all, apart from taking up a lot of CPU time. They could be done better, sure, but they are almost never interacting with anything else anyway so there is little point in trying to improve it right now.

The biggest problem is that "update" is very broad term. There are multiple stages to it, some of which can be threaded, others not so much. And whole update has to happen in that 16.6 ms budget you've mentioned (which doesn't apply to voxel computations or that news download, which makes them easy to deal with). You may have dependencies between these stages, but not within them. As a result, you can go "wide" inside a given stage, splitting into several tasks, then collecting results (or potentially spawning some more tasks depending on results, but eventually they have to meet at one point). Also, if you have multiple stages that don't depend on each other, you can have those execute in parallel as well.

I can't really tell you what these stages are exactly, since that depends on how you decide to look on things, but in general we tend to split update into "Update before simulation", "Simulate", "Update after simulation" and "Draw".

Right now "Simulate" (physics simulation) is closest to being able to run "wide". Since Havok deals with task scheduling on it's own, the only thing preventing us from turning MT physics step on are some issues on our side (unless we switched it on without me noticing :) ). It is possible that whatever is preventing us from turning it on can be fixed without having access to HavokWrapper source and that might give pretty good performance boost in cases where physics simulation is the bottleneck. If you'd like to improve server performance, this might be the lowest hanging fruit.

Remaining stages are currently much harder to run in parallel since they weren't designed with this in mind. Fixing this is really not an issue of threading library, but of the design of each entity. There are also session components (classes derived from MySessionComponent) which tend to have better defined ordering than entities themselves and some of which could potentially by executed in parallel.

**Comment by Drakoskai** [comment]
Gotcha, I'm going read over our conversation a few times.  In the mean time, I think I've replicated the general interface of the the old parallel library and I think my next task should be to see how SE reacts when I replace the old calls with calls to the new one.  I'm hoping that the new library will increase the performance in all cases even if the old pattern isn't taking full advantage of it.  

I just want to emphasis that the new pattern can track dependencies and keep them executing in order over multiple cores -- and you don't have to do it that way.  The most optimal way to use the pattern is to set it up so that multiple producers are streaming events to multiple consumers.

As for work flows, I was talking in general about using the fire a weapon as an example flow, I wasn't implying that SE does it that way.

"It is possible to think of it as a set of tasks, but you seem to lack the bigger picture when deciding what these tasks are. Those voxel geometry tasks are really somewhere on the edge and they barely matter to overall multithreading at all, apart from taking up a lot of CPU time. They could be done better, sure, but they are almost never interacting with anything else anyway so there is little point in trying to improve it right now."

I don't want to get stuck on this too much yet.  I wanted to explore the idea of batching out the calculations either one at a time or in sets since we can make sure they are executed in order over multiple cores which should better load balance the processors, but at this time let's leave it alone.

**Comment by Drakoskai** [comment]
Sorry for the long period of silence, I had to find a new job-- however, at that new job I've been designing and am now deploying a Kafka-storm-hadoop system, hhich has given me some ideas for the dedicated server of space engineers.  I've been digging around in storm's source code and it seems plausible to me that Space Engineers could use the same architectures as in the dedicated server to essentially farm out computational units of work over a cluster of servers.

**Comment by Drakoskai** [comment]
Did an experimental re-write on parallel tasks.  Seems to be working well so far.  there may be a job or two that needs to be re-structured -- also some of the precalc jobs would need to be restructured to load balance better, but this library should do the trick.  I doubt keen will use it, but hey I like to code and I had fun doing it ;p

**Comment by Vestrheim** [comment]
This was an interesting read, where are you with this now Drakoskai, did you get any performance increases ?

**Comment by Drakoskai** [comment]
I've been pulling some crazy hours at work for the last few months, I'm trying to build out all the infrastructure and software for big data processing and machine learning for a company that had no concept of it whatsoever, and they want it now, lol.  I did take a look at some of the things Keen is doing in regard to this subject at it looks like they are working on this are incrementally.  I'm not sure what their vision is or what they've been researching.  To be honest, this isn't going to be a fix where one person does a tweak and boom everything is better.  It's a major refactoring which could have major consequences and Keen doesn't have unit tests to catch issues before release.  I'd like to revisit this, maybe I will once I catch up a little bit on sleep.

**Comment by ProsperCraft** [comment]
I would love to see this, as it is a real struggle to host a public server with the current dedicated server code.
